app/email_batching.go

```diff
@@ -121,24 +121,34 @@ func (job *EmailBatchingJob) handleNewNotifications() {
 }
 
 func (job *EmailBatchingJob) checkPendingNotifications(now time.Time, handler func(string, []*batchedNotification)) {
-	// look for users who've acted since pending posts were received
 	for userId, notifications := range job.pendingNotifications {
-		schan := Srv.Store.Status().Get(userId)
-		pchan := Srv.Store.Preference().Get(userId, model.PREFERENCE_CATEGORY_NOTIFICATIONS, model.PREFERENCE_NAME_EMAIL_INTERVAL)
-		batchStartTime := notifications[0].post.CreateAt
 
-		// check if the user has been active and would've seen any new posts
-		if result := <-schan; result.Err != nil {
-			l4g.Error(utils.T("api.email_batching.check_pending_emails.status.app_error"), result.Err)
-			delete(job.pendingNotifications, userId)
-			continue
-		} else if status := result.Data.(*model.Status); status.LastActivityAt >= batchStartTime {
-			delete(job.pendingNotifications, userId)
-			continue
+		// if the user has viewed any of the channels that we've queued notifications for since the time at which the
+		// first notification was posted, delete all queued notifications so that the email is never sent
+		batchStartTime := notifications[0].post.CreateAt
+		inspectedChannelIds := make(map[string]bool)
+		for _, notification := range notifications {
+			// because multiple messages may have been queued for the same channel, we take care to only check the last viewed at
+			// time of each channel exactly once
+			if !inspectedChannelIds[notification.post.ChannelId] {
+				cchan := Srv.Store.Channel().GetMember(notification.post.ChannelId, userId)
+				if result := <-cchan; result.Err != nil {
+					l4g.Error("Unable to find ChannelMember record", result.Err)
+					delete(job.pendingNotifications, userId)
+					break
+				} else if channelMember := result.Data.(*model.ChannelMember); channelMember.LastViewedAt >= batchStartTime {
```

Comment:
This code is effectively the same as the code that it replaces, except that we're checking the `LastViewedAt` field of the `ChannelMember` struct for the user/channel combination, rather than the `LastActivityAt` field of the user's `Status` struct. This is because the `Status.LastActivityAt` field only seems to be updated when the user's status is set to `online`, but we want this notification discarding code to function even if the user has manually set their status to `away`.